### PR TITLE
[stdlib] SE-0061: Add Generic Result and Error Handling to autoreleas…

### LINKED ISSUE #43454: [SR-842] Add `rethrows` to the `ObjectiveC.autoreleasepool` function
|                  |                 |
   |------------------|-----------------|
   |Previous ID       | SR-842      |
   |Radar             | None         |
   |Original Reporter | @benrimmington      |
   |Type              | Improvement    |
   |Status            | Resolved        |
   |Resolution        | Done    |
        
   <details>
  <summary>Additional Detail from JIRA</summary>

   |                  |                 |
   |------------------|-----------------|
   |Votes             | 0         |
   |Component/s       | Standard Library    |
   |Labels            | Improvement, StarterBug        |
   |Assignee          | @tjw      |
   |Priority          | Medium      |

   

   md5: c5854d2715589b4cb47479682a705263

  </details>

**is duplicated by**:  
* [SR-1394](https://bugs.swift.org/browse/SR-1394) Implement SE-0061 - Add Generic Result and Error Handling to autoreleasepool()  

**Issue Description:**

For example:

``` java
public func autoreleasepool(@noescape body: () throws -> Void) rethrows {
  let pool = __pushAutoreleasePool()
  defer {
    __popAutoreleasePool(pool)
  }
  try body()
}
```

<https://github.com/apple/swift/blob/master/stdlib/public/SDK/ObjectiveC/ObjectiveC.swift>

### PR DESCRIPTION
**Pull request by tjw**
Added `rethrows` and generic return type to `ObjectiveC.autoreleasepool`.
Added tests of these new capabilities.

Fixes #43454

### CODE REVIEW

**Comment by lattner** [comment]
LGTM for functionality.  @gribozavr or @dabrahams can you review for style?

**Comment by tkremenek** [comment]
@swift-ci test

File: test/Interpreter/SDK/autorelease.swift [hunk 1]
```diff
@@ -34,3 +34,39 @@ print("autorelease test end")
 // CHECK-NEXT: object died
 // CHECK-NEXT: after call to useTemp
 // CHECK-NEXT: autorelease test end
+
+// Using an @objc class to check that errors are retained across the pool
+// boundaries. A classic crash is an error created inside a pool and then
+// zombied before handling it outside the pool.
+@objc class Error: NSObject, ErrorProtocol {
```
Comment (gribozavr) [inline test/Interpreter/SDK/autorelease.swift]: Please add a space before the colon.

File: test/Interpreter/SDK/autorelease.swift [hunk 2]
```diff
@@ -34,3 +34,39 @@ print("autorelease test end")
...
+  let message:String
+  init(message:String) {
```
Comment (gribozavr) [inline test/Interpreter/SDK/autorelease.swift]: Please add spaces after colons in variable declarations.

File: test/Interpreter/SDK/autorelease.swift [hunk 3]
```diff
@@ -34,3 +34,39 @@ print("autorelease test end")
...
+    self.message = message
+  }
+}
+
+// Check that rethrow works.
+func requireString(string:String?) throws -> String {
+  guard let string = string else {
+    throw Error(message:"no string")
```
Comment (gribozavr) [inline test/Interpreter/SDK/autorelease.swift]: Spaces after colons.

File: stdlib/public/SDK/ObjectiveC/ObjectiveC.swift [hunk 4]
```diff
@@ -191,10 +191,12 @@ func __pushAutoreleasePool() -> OpaquePointer
 @_silgen_name("_swift_objc_autoreleasePoolPop")
 func __popAutoreleasePool(_ pool: OpaquePointer)
 
-public func autoreleasepool(_ code: @noescape () -> Void) {
+public func autoreleasepool<Result>(_ body: @noescape () throws -> Result) rethrows -> Result {
```
Comment (gribozavr) [inline stdlib/public/SDK/ObjectiveC/ObjectiveC.swift]: Please wrap to 80 columns:

```swift
public func autoreleasepool<Result>(
  _ body: @noescape () throws -> Result
) rethrows -> Result {
```

**Comment by tkremenek** [comment]
@swift-ci test and merge

**Comment by tjw** [comment]
It looks like the Linux build failed for unrelated reasons, if I'm understanding the listed results correctly.

**Comment by tkremenek** [comment]
Indeed.  Re-testing.

**Comment by tkremenek** [comment]
@swift-ci test linux

**Comment by tkremenek** [comment]
@swift-ci please smoke test OS X platform

**Pull request merged**